File: functionapi-api/src/main/java/io/github/cottonmc/functionapi/Registration.java
package io.github.cottonmc.functionapi;

import com.google.common.collect.Lists;
import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import io.github.cottonmc.functionapi.api.content.*;
import io.github.cottonmc.functionapi.api.script.FunctionAPIIdentifier;
import io.github.cottonmc.functionapi.util.FunctionAPIIdentifierImpl;

import java.util.*;
import java.util.function.BiConsumer;
import java.util.stream.Collectors;

@SuppressWarnings("unchecked")
public class Registration implements BiConsumer<CommandFileSource, CommandDispatcher<Map<String,Object>>> {

    private LinkedList<ContentRegistration> contentRegistrations;

    public Registration(){
        contentRegistrations = new LinkedList<>();
        ServiceLoader.load(ContentRegistration.class).iterator().forEachRemaining(contentRegistrations::add);

    }

    private void accept(Map<String,Object> contentRegistrationContext) {
        Map<FunctionAPIIdentifier, ContentRegistrationContext.ContentType> elements = (Map<FunctionAPIIdentifier, ContentRegistrationContext.ContentType>) contentRegistrationContext.get("elements");

        if (elements.isEmpty()) {
            return;
        }

        Map<Integer, List<String>> variants = (Map<Integer, List<String>>) contentRegistrationContext.get("variants");

        if (variants.size() == 0) {
            elements.forEach((identifier, contentType) -> register("", contentRegistrationContext, identifier));
        } else {

            List<List<Object>> permutations = Lists.cartesianProduct(variants.values().toArray(new List[]{}));
            elements.forEach((identifier, contentType) -> {

                for (List<Object> variant : permutations) {
                    String variantName = "_" + variant.stream().map(Object::toString).collect(Collectors.joining("_"));
                   register(variantName,contentRegistrationContext,identifier);
                }
            });
        }
    }

    private void register(String variantName, Map<String, Object> contentRegistrationContext, FunctionAPIIdentifier identifier) {

        identifier = new FunctionAPIIdentifierImpl(identifier.getNamespace(), identifier.getPath() + variantName + contentRegistrationContext.get("postfix"));
        for (ContentRegistration contentRegistration : contentRegistrations) {
            contentRegistration.register(identifier,contentRegistrationContext);
        }

    }


    @Override
    public void accept(CommandFileSource commandFileSource, CommandDispatcher<Map<String,Object>> contentRegistrationContextCommandDispatcher) {
        Map<FunctionAPIIdentifier, List<String>> commands = commandFileSource.getCommandFiles();


        commands.forEach((identifier, list) -> {
            Map<String,Object> context = new HashMap<>();

            if (identifier.getPath().split("/").length == 1) {
                try {
                    for (String command : list) {
                        contentRegistrationContextCommandDispatcher.execute(command, context);
                    }
                } catch (CommandSyntaxException e) {
                    System.err.println(e.getInput());
                    e.printStackTrace();
                }

                accept(context);
            }
        });
    }

}

File: functionapi-content/src/main/java/io/github/cottonmc/functionapi/content/Registration.kt
package io.github.cottonmc.functionapi.content

import com.google.common.collect.Lists
import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.exceptions.CommandSyntaxException
import io.github.cottonmc.functionapi.api.FunctionAPIIdentifier
import io.github.cottonmc.functionapi.api.content.ContentRegistration
import io.github.cottonmc.functionapi.api.content.registration.RegistrationContainer
import io.github.cottonmc.functionapi.content.template.RegistrationTemplate
import io.github.cottonmc.functionapi.util.documentation.ContentCommandDocumentationGenerator
import io.github.cottonmc.functionapi.util.impl.FunctionAPIIdentifierImpl
import io.github.cottonmc.functionapi.util.MissingResourceException
import io.github.cottonmc.functionapi.util.impl.PermanentHashmap
import java.io.Writer
import java.util.ArrayList
import java.util.function.Consumer
import java.util.stream.Collectors

class Registration(private val subfolder: String,
                   private var documentationGenerator: ContentCommandDocumentationGenerator = ContentCommandDocumentationGenerator(subfolder)
) {

    constructor(subfolder: String,writer: Writer):this(subfolder,ContentCommandDocumentationGenerator(writer))
    private var context = PermanentHashmap<String, Any>()
    val staticCommandExecutor = StaticCommandExecutor()

    fun register(contentRegistrations: List<ContentRegistration>): CommandDispatcher<Map<String, Any>> {

        context = PermanentHashmap()

        documentationGenerator.generate(staticCommandExecutor.commandDispatcher)

        val commands: Set<FunctionAPIIdentifier> = staticCommandExecutor.iDs
        commands.forEach(Consumer { identifier: FunctionAPIIdentifier ->
            if (identifier.path.startsWith("$subfolder/") && identifier.path.split("/").toTypedArray().size == 2 && identifier.path.endsWith(".mccontent")) {
                try {
                    staticCommandExecutor.execute(FunctionAPIIdentifierImpl(identifier.namespace, identifier.path), context)
                } catch (e: CommandSyntaxException) {
                    e.printStackTrace()
                } catch (e: MissingResourceException) {
                    e.printStackTrace()
                }
                generateIdsAndRegister(contentRegistrations, context)
            }
        })

        return staticCommandExecutor.commandDispatcher
    }

    private fun generateIdsAndRegister(contentRegistrations: List<ContentRegistration>, context: Map<String, Any>) {
        contentRegistrations.forEach { it.setup(context) }
        val registrationContainer = context.getOrDefault("registration",RegistrationTemplate()) as RegistrationContainer

        val ids = registrationContainer.getIDs()
        var postfix = (registrationContainer.getPostfixes()).stream().collect(Collectors.joining("_"))
        if (!postfix.isEmpty()) {
            postfix = "_$postfix"
        }
        val variants = (registrationContainer.getVariants())
        val variantNames: Collection<List<String>> = Lists.cartesianProduct(ArrayList(variants.values))

        if (!variantNames.isEmpty()) {
            for (variantName in variantNames) {
                var variant = variantName.stream().collect(Collectors.joining("_"))
                variant += postfix
                for (functionAPIIdentifier in ids) {
                    contentRegistrations.forEach { it.register(FunctionAPIIdentifierImpl(functionAPIIdentifier.namespace, "${functionAPIIdentifier.path}_$variant")) }
                }
            }
        } else {
            for (functionAPIIdentifier in ids) {
                contentRegistrations.forEach { it.register(functionAPIIdentifier) }
            }
        }
    }
}